zulip/zulip/__init__.py

```diff
@@ -697,24 +697,41 @@ def call_endpoint(
         method: str = "POST",
         request: Optional[Dict[str, Any]] = None,
         longpolling: bool = False,
+        retry_on_rate_limit_error: bool = False,
         files: Optional[List[IO[Any]]] = None,
         timeout: Optional[float] = None,
     ) -> Dict[str, Any]:
+        secs: float
+
         if request is None:
             request = dict()
         marshalled_request = {}
         for (k, v) in request.items():
             if v is not None:
                 marshalled_request[k] = v
         versioned_url = API_VERSTRING + (url if url is not None else "")
-        return self.do_api_query(
-            marshalled_request,
-            versioned_url,
-            method=method,
-            longpolling=longpolling,
-            files=files,
-            timeout=timeout,
-        )
+
+        while True:
+            result = self.do_api_query(
+                marshalled_request,
+                versioned_url,
+                method=method,
+                longpolling=longpolling,
+                files=files,
+                timeout=timeout,
+            )
+            if not retry_on_rate_limit_error or result["result"] == "success":
+                break
```

Comment:
I'd do this as two conditionals, first the `success` one, and then the other one.  We like to do this so that code coverage tools can tell you whether we have tests for each independent condition.